### PR TITLE
JWT Claim constraints

### PR DESCRIPTION
**Pull request by bartv**
# Description

This PR implements custom claim support.

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/config.py [hunk 1]
```diff
@@ -445,6 +447,80 @@ def __init__(self, name: str, port: int = 8888) -> None:
 # auth
 #############################
 AUTH_JWT_PREFIX = "auth_jwt_"
+AUTH_JWT_CLAIM_RE = r"^\s*([^\s].+)\s+(in|is)\s+([^\s].+)\s*$"
+
+claim_type = dict[str, str | list[str]]
+
+
+@dataclasses.dataclass
```
Comment (wouterdb) [inline src/inmanta/config.py:455]: All of the auth config could be moved into a separate module or in the server config?

Config is loaded very early, so we do have to be aware for import loops.

**Review by wouterdb — COMMENTED** [review]
It would be interesting if we could test it some more completely. Perhaps generating a token ( we have that in testing already) and make a client call would be interesting (with e.g. valid, invalid and garbage claims).

File: src/inmanta/config.py [hunk 2]
```diff
@@ -445,6 +447,80 @@ def __init__(self, name: str, port: int = 8888) -> None:
...
+    :return: The result of the check
+    """
+    try:
+        return all(constraint.match_claim(claims) for constraint in claim_constraints)
+    except Exception as e:
+        LOGGER.info(f"The configured claim constraints failed to evaluate against the provided claims: {e}")
+        return False
 
 
 class AuthJWTConfig:
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/config.py:526]: But is not a problem I want to solve in this ticket ;-)

File: src/inmanta/config.py (hunk 1 above)
Comment (bartv) [reply src/inmanta/config.py]: seems to work

**Comment by bartv** [comment]
> Looks nice
> 
> It would be interesting if we could test it some more completely. Perhaps generating a token ( we have that in testing already) and make a client call would be interesting (with e.g. valid, invalid and garbage claims).

done

File: src/inmanta/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/config.py:452]: Fair.

File: src/inmanta/config.py (hunk 1 above)
Comment (bartv) [reply src/inmanta/config.py]: Not sure how I will be able to move this into inmanta.protocol without import problems. See CI

**Review by wouterdb — APPROVED** [review]

**Review by wouterdb — APPROVED** [review]

**Pull request closed without merging**